lucene/core/src/java/org/apache/lucene/search/AbstractKnnVectorQuery.java

```diff
@@ -156,6 +196,58 @@ private TopDocs getLeafResults(
     }
   }
 
+  private DocIdSetIterator executeSeedQuery(LeafReaderContext ctx, Weight seedWeight)
+      throws IOException {
+    if (seedWeight != null) {
+      // Execute the seed query
+      TopScoreDocCollector seedCollector =
+          new TopScoreDocCollectorManager(
+                  k /* numHits */,
+                  null /* after */,
+                  Integer.MAX_VALUE /* totalHitsThreshold */,
+                  false /* supportsConcurrency */)
+              .newCollector();
+      LeafCollector leafCollector;
+      try {
+        leafCollector = seedCollector.getLeafCollector(ctx);
+      } catch (
+          @SuppressWarnings("unused")
+          CollectionTerminatedException e) {
+        // there is no doc of interest in this reader context
+        // continue with the following leaf
+        leafCollector = null;
+      }
+      if (leafCollector != null) {
+        BulkScorer scorer = seedWeight.bulkScorer(ctx);
+        if (scorer != null) {
+          try {
+            scorer.score(leafCollector, ctx.reader().getLiveDocs());
+          } catch (
+              @SuppressWarnings("unused")
+              CollectionTerminatedException e) {
+            // collection was terminated prematurely
+            // continue with the following leaf
+          }
+        }
+        leafCollector.finish();
+      }
+
+      TopDocs seedTopDocs = seedCollector.topDocs();
+      return convertDocIdsToVectorOrdinals(ctx, new TopDocsDISI(seedTopDocs));
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Returns a new iterator that maps the provided docIds to the vector ordinals.
+   *
+   * @lucene.internal
+   * @lucene.experimental
+   */
+  protected abstract DocIdSetIterator convertDocIdsToVectorOrdinals(
+      LeafReaderContext ctx, DocIdSetIterator docIds) throws IOException;
```

Comment:
Since both implementations do a `ctx.reader()` call, wondering if the signature might be narrower then?
```suggestion
      LeafReader reader, DocIdSetIterator docIds) throws IOException;
```